Code/GraphMol/SubstructLibrary/SubstructLibrary.cpp

```diff
@@ -126,144 +123,156 @@ void SubSearcher(const ROMol &in_query, const Bits &bits,
-    fut.get();
+#ifdef RDK_THREADSAFE_SSS
+  if (numThreads > 1) {
+    std::vector<int> counterVect(numThreads, 0);
+    int maxResultsPerThread = maxResults;
+    if (maxResults != -1) {
+      maxResultsPerThread /= numThreads;
+    }
+    std::vector<int> maxResultsVect(numThreads, maxResultsPerThread);
+    std::vector<unsigned int> endIdxVect(numThreads, endIdx);
+    if (maxResults != -1) {
+      int excess = maxResults % numThreads;
+      for (int i = 0; i < excess; ++i) {
+        ++maxResultsVect[i];
+      }
+    }
+    std::vector<std::future<void>> thread_group;
+    std::vector<std::vector<unsigned int>> internal_results;
+    if (idxs) {
+      internal_results.resize(numThreads);
+    }
+    int thread_group_idx;
+    for (thread_group_idx = 0; thread_group_idx < numThreads;
+         ++thread_group_idx) {
+      // need to use boost::ref otherwise things are passed by value
+      thread_group.emplace_back(
+          std::async(std::launch::async, SubSearcher, std::ref(query), bits,
+                     std::ref(mols), startIdx + thread_group_idx,
+                     std::ref(endIdxVect[thread_group_idx]), numThreads,
+                     needs_rings, std::ref(counterVect[thread_group_idx]),
+                     maxResultsVect[thread_group_idx],
+                     idxs ? &internal_results[thread_group_idx] : nullptr));
+    }
+    if (maxResults != -1) {
+      // If we are running with maxResults in a multi-threaded settings,
+      // some threads may have screened more molecules than others.
+      // If maxResults was close to the theoretical maximum, some threads
+      // might have even run out of molecules to screen without reaching
+      // maxResults so we need to make sure that all threads have screened as
+      // many molecules as the most productive thread if we want multi-threaded
+      // runs to yield the same results independently from the number of
+      // threads.
+      thread_group_idx = 0;
+      for (auto &fut : thread_group) {
+        fut.get();
+        counter += counterVect[thread_group_idx++];
+      }
+      thread_group.clear();
+      // Find out out the max number of molecules that was screened by the most
+      // productive thread and do the same in all other threads, unless the
+      // max number of molecules was reached
+      unsigned int maxEndIdx =
+          *std::max_element(endIdxVect.begin(), endIdxVect.end());
+      for (thread_group_idx = 0; thread_group_idx < numThreads;
+           ++thread_group_idx) {
+        if (endIdxVect[thread_group_idx] >= maxEndIdx) {
+          continue;
+        }
+        // need to use boost::ref otherwise things are passed by value
+        thread_group.emplace_back(std::async(
+            std::launch::async, SubSearcher, std::ref(query), bits,
+            std::ref(mols), endIdxVect[thread_group_idx] + numThreads,
+            std::ref(maxEndIdx), numThreads, needs_rings,
+            std::ref(counterVect[thread_group_idx]), -1,
```

Comment:
It might be more elegant to have the thread return the # screened/ # matched via

```
  std::future<unsigned int> num_matched = fut.get();
```

Reply:
Or maybe std::pair